src/commands/trigger.cpp

```
@@ -0,0 +1,728 @@
+  types_summary[AFTER_INSERT_ROW] |= TRIGGER_TYPE_MATCHES(
+    type, TRIGGER_TYPE_ROW, TRIGGER_TYPE_AFTER, TRIGGER_TYPE_INSERT);
+  types_summary[AFTER_INSERT_STATEMENT] |= TRIGGER_TYPE_MATCHES(
+    type, TRIGGER_TYPE_STATEMENT, TRIGGER_TYPE_AFTER, TRIGGER_TYPE_INSERT);
+  types_summary[AFTER_UPDATE_ROW] |= TRIGGER_TYPE_MATCHES(
+    type, TRIGGER_TYPE_ROW, TRIGGER_TYPE_AFTER, TRIGGER_TYPE_UPDATE);
+  types_summary[AFTER_UPDATE_STATEMENT] |= TRIGGER_TYPE_MATCHES(
+    type, TRIGGER_TYPE_STATEMENT, TRIGGER_TYPE_AFTER, TRIGGER_TYPE_UPDATE);
+  types_summary[AFTER_DELETE_ROW] |= TRIGGER_TYPE_MATCHES(
+    type, TRIGGER_TYPE_ROW, TRIGGER_TYPE_AFTER, TRIGGER_TYPE_DELETE);
+  types_summary[AFTER_DELETE_STATEMENT] |= TRIGGER_TYPE_MATCHES(
+    type, TRIGGER_TYPE_STATEMENT, TRIGGER_TYPE_AFTER, TRIGGER_TYPE_DELETE);
+}
+
+/**
+ * execute trigger on each row before inserting tuple.
+ */
+storage::Tuple *TriggerList::ExecBRInsertTriggers(storage::Tuple *tuple, executor::ExecutorContext *executor_context_) {
+  unsigned i;
+  LOG_DEBUG("enter into ExecBRInsertTriggers");
+
+  //check valid type
+  if (!types_summary[EnumTriggerType::BEFORE_INSERT_ROW]) {
+    return nullptr;
+  }
+
+  storage::Tuple *new_tuple = tuple;
+  for (i = 0; i < triggers.size(); i++) {
+    Trigger &obj = triggers[i];
+    int16_t trigger_type = obj.GetTriggerType();
+    //check valid type
+    if (!TRIGGER_TYPE_MATCHES(trigger_type, TRIGGER_TYPE_ROW, TRIGGER_TYPE_BEFORE, TRIGGER_TYPE_INSERT)) {
+      continue;
+    }
+
+    //TODO: check if trigger is enabled
+
+    expression::AbstractExpression *predicate_ = obj.GetTriggerWhen();
+    if (predicate_ != nullptr) {
+      LOG_DEBUG("predicate_ is not nullptr");
+      LOG_DEBUG("predicate_ type = %s", ExpressionTypeToString(predicate_->GetExpressionType()).c_str());
+      LOG_DEBUG("predicate_ size = %lu", predicate_->GetChildrenSize());
+      if (executor_context_ != nullptr) {
```

I feel like this should never be null. We might want to throw an error instead.